src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/CharSetSolver.cs

```diff
@@ -2,327 +2,407 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.Globalization;
+using System.Runtime.InteropServices;
+using System.Threading;
 
 namespace System.Text.RegularExpressions.Symbolic
 {
     /// <summary>
-    /// Provides functionality to build character sets, to perform boolean operations over character sets,
-    /// and to construct a symbolic finite automata (SFA) over character sets from a regex.
-    /// Character sets are represented by bitvector sets.
+    /// Provides functionality to build character sets represented as <see cref="BDD"/>s
+    /// and to perform boolean operations over those sets.
     /// </summary>
-    internal sealed class CharSetSolver : BDDAlgebra, ICharAlgebra<BDD>
+    internal sealed class CharSetSolver : ISolver<BDD>
     {
-        /// <summary>BDDs for all ASCII characters for fast lookup.</summary>
-        private readonly BDD[] _charPredTable = new BDD[128];
-        internal readonly BDD _nonAscii;
-
-        /// <summary>Initialize the solver.</summary>
-        /// <remarks>Consumers should use the singleton <see cref="Instance"/>.</remarks>
-        private CharSetSolver()
-        {
-            _nonAscii = CreateCharSetFromRange('\x80', '\uFFFF');
-        }
-
-        /// <summary>Singleton instance of <see cref="CharSetSolver"/>.</summary>
-        public static CharSetSolver Instance { get; } = new CharSetSolver();
-
-        /// <summary>
-        /// Make a character predicate for the given character c.
-        /// </summary>
-        public BDD CharConstraint(char c)
-        {
-            // individual character BDDs are always fixed
-            BDD[] charPredTable = _charPredTable;
-            return c < charPredTable.Length ?
-                charPredTable[c] ??= CreateBDDFromChar(c) :
-                CreateBDDFromChar(c);
-        }
-
-        private BDD CreateBDDFromChar(ushort c)
+        /// <summary>BDD for each ASCII character that returns true for that one character.</summary>
+        /// <remarks>This cache is shared amongst all CharSetSolver instances and is accessed in a thread-safe manner.</remarks>
+        private static readonly BDD?[] s_asciiCache = new BDD[128];
+        /// <summary>BDD that returns true for every non-ASCII character.</summary>
+        /// <remarks>This instance is shared amongst all CharSetSolver instances and is accessed in a thread-safe manner.</remarks>
+        private static BDD? s_nonAscii;
```

Comment:
Does our existing test coverage guarantee that these will be accessed in a thread-safe manner? Or asked differently, should we expect something to break in our tests of the way these are accessed is ever changed incorrectly?

Reply:
oh NVM, I see that we just use Interlocked.CompareExchange.